File: java/src/test/java/io/github/webictbyleo/odoid/PerformanceTest.java
package io.github.webictbyleo.odoid;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Performance tests — SPEC.md § 9.
 * Each encode / decode call must average ≤ 0.1000 ms.
 */
class PerformanceTest {

    private static final double LIMIT_MS  = 0.1;
    private static final int    WARMUP    = 2_000;
    private static final int    ITERATIONS = 10_000;

    private void assertAvgMs(String label, Runnable fn) {
        for (int i = 0; i < WARMUP; i++) fn.run();

        long start = System.nanoTime();
        for (int i = 0; i < ITERATIONS; i++) fn.run();
        double avgMs = (System.nanoTime() - start) / 1_000_000.0 / ITERATIONS;

        assertTrue(avgMs <= LIMIT_MS,
            String.format("%s: average %.4f ms exceeded spec limit of %.4f ms", label, avgMs, LIMIT_MS));
    }

    @Test
    void encode_length6_averageBelowLimit() {
        long[] n = {0};
        assertAvgMs("encode/6", () -> {
            n[0] = (n[0] + 1) % Charsets.MAX.get(6);
            OdoId.encode(n[0], 6);
        });
    }

    @Test
    void encode_length7_averageBelowLimit() {
        long[] n = {0};
        assertAvgMs("encode/7", () -> {
            n[0] = (n[0] + 1) % Charsets.MAX.get(7);
            OdoId.encode(n[0], 7);
        });
    }

    @Test
    void encode_length8_averageBelowLimit() {
        long[] n = {0};
        assertAvgMs("encode/8", () -> {
            n[0] = (n[0] + 1) % Charsets.MAX.get(8);
            OdoId.encode(n[0], 8);
        });
    }

    @Test
    void decode_length6_averageBelowLimit() {
        String[] ids = {"0A0000", "0D7NM7", "ZZ9ZZZ", "1B3C4D", "AB0000"};
        int[] i = {0};
        assertAvgMs("decode/6", () -> OdoId.decode(ids[i[0]++ % ids.length]));
    }

    @Test
    void decode_length7_averageBelowLimit() {
        String[] ids = {"0A00000", "0A15NM7", "ZZ9ZZZZ", "1B3C4D5", "AB00000"};
        int[] i = {0};
        assertAvgMs("decode/7", () -> OdoId.decode(ids[i[0]++ % ids.length]));
    }

    @Test
    void decode_length8_averageBelowLimit() {
        String[] ids = {"0A000000", "ZZ9ZZZZZ", "1B3C4D5E", "AB000000"};
        int[] i = {0};
        assertAvgMs("decode/8", () -> OdoId.decode(ids[i[0]++ % ids.length]));
    }
}

File: go/odoid/performance_test.go
// Performance tests — SPEC.md § 9.
// Each encode() / decode() call must average ≤ 0.1000 ms over a representative load.
//
// Run benchmarks with: go test -bench=. -benchtime=5s ./odoid/
package odoid_test

import (
	"fmt"
	"testing"
	"time"

	"github.com/Webictbyleo/odoid/go/odoid"
)

const (
	perfLimitMs = 0.1
	perfWarmup  = 2_000
	perfIter    = 10_000
)

func assertAvgMs(t *testing.T, label string, fn func()) {
	t.Helper()
	for i := 0; i < perfWarmup; i++ {
		fn()
	}
	start := time.Now()
	for i := 0; i < perfIter; i++ {
		fn()
	}
	avgMs := float64(time.Since(start).Nanoseconds()) / float64(perfIter) / 1e6
	fmt.Printf("RESULT|go|%s|%.6f\n", label, avgMs)
	if avgMs > perfLimitMs {
		t.Errorf("%s: average %.4f ms exceeded spec limit of %.4f ms", label, avgMs, perfLimitMs)
	}
}

func TestPerformanceEncodeLength6(t *testing.T) {
	n := uint64(0)
	assertAvgMs(t, "encode/6", func() {
		n = (n + 1) % odoid.Max[6]
		odoid.Encode(n, 6) //nolint:errcheck
	})
}

func TestPerformanceEncodeLength7(t *testing.T) {
	n := uint64(0)
	assertAvgMs(t, "encode/7", func() {
		n = (n + 1) % odoid.Max[7]
		odoid.Encode(n, 7) //nolint:errcheck
	})
}

func TestPerformanceEncodeLength8(t *testing.T) {
	n := uint64(0)
	assertAvgMs(t, "encode/8", func() {
		n = (n + 1) % odoid.Max[8]
		odoid.Encode(n, 8) //nolint:errcheck
	})
}

func TestPerformanceDecodeLength6(t *testing.T) {
	ids := []string{"0A0000", "0D7NM7", "ZZ9ZZZ", "1B3C4D", "AB0000"}
	i := 0
	assertAvgMs(t, "decode/6", func() {
		odoid.Decode(ids[i%len(ids)]) //nolint:errcheck
		i++
	})
}

func TestPerformanceDecodeLength7(t *testing.T) {
	ids := []string{"0A00000", "0A15NM7", "ZZ9ZZZZ", "1B3C4D5", "AB00000"}
	i := 0
	assertAvgMs(t, "decode/7", func() {
		odoid.Decode(ids[i%len(ids)]) //nolint:errcheck
		i++
	})
}

func TestPerformanceDecodeLength8(t *testing.T) {
	ids := []string{"0A000000", "ZZ9ZZZZZ", "1B3C4D5E", "AB000000"}
	i := 0
	assertAvgMs(t, "decode/8", func() {
		odoid.Decode(ids[i%len(ids)]) //nolint:errcheck
		i++
	})
}

func TestPerformanceGeneratorNext(t *testing.T) {
	g, _ := odoid.NewOdoIDGenerator(odoid.GeneratorConfig{Length: 6})
	assertAvgMs(t, "generate/6", func() {
		g.Next() //nolint:errcheck
	})
}

// ── Benchmarks (go test -bench=. -benchtime=5s ./odoid/) ──────────────────

func BenchmarkEncode6(b *testing.B) {
	n := uint64(0)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		n = (n + 1) % odoid.Max[6]
		odoid.Encode(n, 6) //nolint:errcheck
	}
}

func BenchmarkEncode7(b *testing.B) {
	n := uint64(0)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		n = (n + 1) % odoid.Max[7]
		odoid.Encode(n, 7) //nolint:errcheck
	}
}

func BenchmarkEncode8(b *testing.B) {
	n := uint64(0)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		n = (n + 1) % odoid.Max[8]
		odoid.Encode(n, 8) //nolint:errcheck
	}
}

func BenchmarkDecode6(b *testing.B) {
	ids := []string{"0A0000", "0D7NM7", "ZZ9ZZZ", "1B3C4D", "AB0000"}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		odoid.Decode(ids[i%len(ids)]) //nolint:errcheck
	}
}

func BenchmarkDecode8(b *testing.B) {
	ids := []string{"0A000000", "ZZ9ZZZZZ", "1B3C4D5E", "AB000000"}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		odoid.Decode(ids[i%len(ids)]) //nolint:errcheck
	}
}
